Use pyproject.toml with hatch

getzze opened the pull request:
Modernize the building and testing framework:

* use pyproject.toml
* use hatch environments
* ~~add test_script to test plotting figures~~ -> added in #155 
* untrack `docs/build`

To build: `hatch build` or `python -m build`
To bump version: `hatch version minor`
To test: `hatch run tests:run`
To test with coverage: `hatch run tests:run-cov && hatch run coverage:run`
To build and test docs: `hatch run docs:all`
To run examples: `hatch run examples:run -- --plot_type=boxplot --data_type=df_with_group_and_hue`


To test #155 I had to write the `test_script.py` file and it was very convenient to use `hatch` environments for testing. So I thought I should actually make a PR about it :)

getzze commented:
I modified a bit the Github action so instead of calling hatch once to run all the tests, the workflow creates a matrix of python and seaborn versions (we could also add an OS matrix if needed, see https://github.com/platformdirs/platformdirs/ for the modifications to do).

Github runs matrix jobs in parallel so the tests are substantially faster.

getzze commented:
I forgot to mention that python 3.7 is not tested (hatch was not working for me with this python version). It would be better to drop it, as it is EOL anyway.

getzze commented:
The notebook runs correctly with python 3.12.

I added support for python 3.13 (only compatible with seaborn 0.13), the tests are passing.

getzze commented:
@trevismd it's ready for review

trevismd commented on pyproject.toml:
```diff
@@ -9,15 +9,14 @@ name = "statannotations"
 readme = "README.md"
 license = {text = "MIT License"}
-requires-python = ">=3.7"
+requires-python = ">=3.8"
```
Ah forgot about this one... Do you mind if I cherry-pick this one independently?

getzze replied:
Yes, sure!
If you are planning a patch release, I realized the README also needs minor corrections: mention that the last version is 0.6, should be 0.7.

getzze commented:
Yes, sure!
If you are planning a patch release, I realized the README also needs minor corrections: mention that the last version is 0.6, should be 0.7.

On 19 November 2024 18:26:45 GMT+00:00, Florian Charlier ***@***.***> wrote:
>@trevismd commented on this pull request.
>
>
>
>> @@ -9,15 +9,14 @@ name = "statannotations"
> description = "add statistical significance or custom annotations on seaborn plots. Based on statannot 0.2.3"
> readme = "README.md"
> license = {text = "MIT License"}
>-requires-python = ">=3.7"
>+requires-python = ">=3.8"
>
>Ah forgot about this one... Do you mind if I cherry-pick this one independently?
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/trevismd/statannotations/pull/156#pullrequestreview-2446277389
>You are receiving this because you authored the thread.
>
>Message ID: ***@***.***>

getzze commented:
closes #150 

I moved all the source files to a `src` folder, which avoids some errors when running tests and building docs.

Also I added some linting scripts: `typos` , end-of-file-fixer and trailing-whitespace

getzze commented:
Hi @trevismd , 
I rebased this PR. It seems it has a lot of changes but most of it is getting rid of the `build` docs folder (that should not be tracked with git).
I believe this PR will help make the project more maintainable by moving to modern packaging with `pyproject.toml`.

The next step would be to add type annotations and a fast linter (`ruff`). But for that we need to have a `pyproject.toml`.

A reverted moving the source to a `src` folder as it is not critical and it makes the changes less trackable. If needed it can be done in a separate PR that does only that.

phaleanopsis commented:
I think it would make sense to merge this PR, and then as a next step work to improve docstrings and type annotations to make the package and documentation a bit more accessible?

getzze committed:
add pyproject.toml and hatch.toml

getzze committed:
remove docs/build

getzze committed:
add pre-commit

getzze committed:
update .gitignore

getzze committed:
update pyproject.toml

getzze committed:
remove unused files

getzze committed:
add docs/sourc/_static

getzze committed:
update github actions

getzze committed:
update python versions

getzze committed:
fix tests and docs

getzze committed:
fix some tests

getzze committed:
fix typo

getzze committed:
fix docs warning

getzze committed:
fix tests failing with pandas>=3

getzze committed:
rework github action test matrix

getzze commented:
I rebased this PR, bumping the minimal python version to 3.10.

getzze committed:
add zizmor to check github actions security issues

getzze committed:
tests working with pandas>=3

getzze committed:
get full coverage

getzze commented:
@trevismd I fixed some small things (to get the test working with pandas>=3 and getting full coverage).
It's ready for review!
